src/mlpack/methods/neighbor_search/spill_search_impl.hpp

```diff
@@ -0,0 +1,251 @@
+    const MetricType metric) :
+    neighborSearch(naive, singleMode, epsilon, metric),
+    tau(tau),
+    rho(rho),
+    leafSize(leafSize)
+{
+  if (tau < 0)
+    throw std::invalid_argument("tau must be non-negative");
+  if (rho < 0 || rho > 1)
+    throw std::invalid_argument("rho must be in the range [0,1]");
+}
+
+// Clean memory.
+template<typename MetricType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType>
+SpillSearch<MetricType, MatType, HyperplaneType, SplitType>::
+~SpillSearch()
+{
+  /* Nothing to do */
+}
+
+template<typename MetricType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType>
+void SpillSearch<MetricType, MatType, HyperplaneType, SplitType>::
+Train(const MatType& referenceSet)
+{
+  if (Naive())
+    neighborSearch.Train(referenceSet);
+  else
+  {
+    // Build reference tree with proper value for tau.
+    Tree* tree = new Tree(referenceSet, tau, leafSize, rho);
+    neighborSearch.Train(tree);
+    // Give the model ownership of the tree.
+    neighborSearch.treeOwner = true;
+  }
+}
+
+template<typename MetricType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType>
+void SpillSearch<MetricType, MatType, HyperplaneType, SplitType>::
+Train(MatType&& referenceSetIn)
+{
+  if (Naive())
+    neighborSearch.Train(std::move(referenceSetIn));
+  else
+  {
+    // Build reference tree with proper value for tau.
+    Tree* tree = new Tree(std::move(referenceSetIn), tau, leafSize, rho);
+    neighborSearch.Train(tree);
+    // Give the model ownership of the tree.
+    neighborSearch.treeOwner = true;
+  }
+}
+
+template<typename MetricType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType>
+void SpillSearch<MetricType, MatType, HyperplaneType, SplitType>::
+Train(Tree* referenceTree)
+{
+  neighborSearch.Train(referenceTree);
+}
+
+template<typename MetricType,
+         typename MatType,
+         template<typename HyperplaneMetricType> class HyperplaneType,
+         template<typename SplitBoundT, typename SplitMatT> class SplitType>
+void SpillSearch<MetricType, MatType, HyperplaneType, SplitType>::
+Search(const MatType& querySet,
+       const size_t k,
+       arma::Mat<size_t>& neighbors,
+       arma::mat& distances)
+{
+  if (Naive() || SingleMode())
+    neighborSearch.Search(querySet, k, neighbors, distances);
+  else
+  {
+    // For Dual Tree Search on SpillTrees, the queryTree must be built with non
+    // overlapping (tau = 0).
+    Tree queryTree(querySet, 0 /* tau */, leafSize, rho);
```

Comment:
I have removed the specialization of `NeighborSearchRules` for `SpillTrees`, so we don't have much duplicated code now.